mautic/mautic
# Stop duplicate campaign update

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [x]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [?]
| Automated tests included?              | [x] 
| Related user documentation PR URL      | /
| Related developer documentation PR URL | /
| Issue(s) addressed                     | /



#### Description:

This PR decouples the segment member updates from the campaign member updates.

Removing the campaign membership build from segment membership will help with performance. The command `bin/console mautic:segments:update` will no longer update campaign membership which it shouldn't be doing anyway. Instead, the command `bin/console mautic:campaigns:update` will take care of the campaign members. Both commands are mandatory for Mautic to work, so there is no harm in removing campaign membership build from the segment membership build command.

This PR also optimizes how the batch segment change works as it was executing both adding and removing operations no matter if any segments were configured for both.

#### Steps to test this PR:
1. Load up [this PR](https://mautibox.com)
2. Create a segment with a filter that will match some contacts
3. Execute `bin/console mautic:segments:update` if you don't have cron configured on your testing instance.
4. Create a campaign that will add some points to all contacts. Use the previous segment as the campaign source.
5. Execute `bin/console mautic:campaigns:update`
6. Ensure that all contacts from the segment are members of the campaign.
7. Add a contact to the segment and remove another contact from the segment to test both options. Doing this manually from the contact list page as a batch operation is the simplest way to do it, but feel free to do it another way.
8. Execute `bin/console mautic:segments:update`. This WILL NOT change the campaign membership.
9. Execute `bin/console mautic:campaigns:update`. This WILL update the campaign membership to match the segment membership.

### BC changes
`Mautic\CampaignBundle\EventListener\LeadSubscriber::onLeadListBatchChange()` and `Mautic\CampaignBundle\EventListener\LeadSubscriber:: onLeadListChange()` public methods are removed, but I cannot imagine someone would extend a subscriber methods so I don't consider this a BC break. Just want to mention this if anyone has different opinion.


<a href="https://gitpod.io/#https://github.com/mautic/mautic/pull/10550"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>
